### ISSUE TITLE
Migration to RTK, unable to have type safe reducers

### ISSUE THREAD

**Issue by MikelArnaiz** [reduxjs/redux-toolkit#2243]
I'm in the process of migrating a codebase to Redux Toolkit and I failed as early as I moved from `createStore` to `configureStore`.

The isse is that the reducers' actions are typed. I'm aware that typing actions it's somehow a lie because reducers are called with every dispatched action + special actions like `@@INIT`, but then how do you maintain type safety/intelligence in reducers?

[According to @markerikson](https://stackoverflow.com/a/63904925), the answer is to not use manual reducers and use `createSlice`. I'd like to do so but I want the migration to be gradual. Once I have the store set up I can start rewriting reducers to slices.

```ts
const FOO_ACTION_1 = 'FOO_ACTION_1';
type FooAction1 = { type: typeof FOO_ACTION_1; payload: number };
type FooAction = FooAction1;
type FooState = { value: number };
const initialFooState: FooState = { value: 0 };

const reducerFoo = (state = initialFooState, action: FooAction): FooState => {
  switch (action.type) {
    case FOO_ACTION_1:
      return { ...state, value: state.value + action.payload };
    default: {
      return state;
    }
  }
};

export const store = configureStore({
  reducer: {
    foo: reducerFoo, // Error
  },
});
```

Which throws

```ts
Type '(state: FooState | undefined, action: FooAction1) => FooState' is not assignable to type 'Reducer<FooState, AnyAction>'.
  Types of parameters 'action' and 'action' are incompatible.
    Property 'payload' is missing in type 'AnyAction' but required in type 'FooAction1'.ts(2322)
store.ts(4, 48): 'payload' is declared here.
```

Reproducible at https://codesandbox.io/s/redux-toolkit-error-typed-actions-hlnbl8.

I could indeed fix it by typing `action: AnyAction` but then `payload` isn't typed.
Alternatively I could provide generics to `configureStore`:

```ts
export const store = configureStore<
  {
    foo: FooState;
  },
  FooAction
>({
  reducer: {
    foo: reducerFoo,
  },
});
```

But I wonder if this is right and scalable, it works in this dummy example but my real project has 30 reducers. Furthermore, you are suppose to derive the app state from the store, e.g: `type AppState = ReturnType<typeof store.getState>`, no provide it to it.

To sum up, I'd like to know how to have an _easy_ migration while maintaining type safety.

**Comment by markerikson** [comment]
Can you clarify what you mean by "type safety" here, specifically?

One bit of confusion we see is that a number of people try to limit _which_ actions can be dispatched to the store at all at the type level.  We see that as an anti-pattern, and specifically discourage it:

https://redux.js.org/usage/usage-with-typescript#avoid-action-type-unions

As far as typing this individual reducer: yeah, I think the type mismatch here is that you're telling TS "this reducer _only_ accepts actions which have a specific `payload` field", but that's stricter than the `Reducer` type that declares "this reducer can (and _will_) be run with any action at all, so it really only has to have a `type" field.

As you noted, the "right" answer here is "convert this reducer to `createSlice` ASAP" :)  But I realize that's not always feasible due to time constraints.  (I'm dealing with that problem myself at my day job at https:replay.io :)  )

As a stopgap, I think your best option here as an in-between stopgap is to write TS type guard functions and use those as case statements:

```ts
const isFooAction = (action: AnyAction) : action is FooAction => {
  return [ACTION_TYPE_1, ACTION_TYPE_2].includes(action.type);
}

function fooReducer(state = initialState, action: AnyAction) {
  if (isFooAction(action)) {
    switch(action.type) {
      // now the cases and types should line up exactly
    }
  }
  return state;
}
```

**Comment by MikelArnaiz** [comment]
Nice read thanks.

With type safe I mainly meant being able to get the payload type. With the guard I'm able to narrow down the action and get the correct payload type. But I see a downside, you get _the risk_ or not keeping the action types array in sync with the action union. We need to compromise and I guess this is still the right thing to do while migrating to slices.

```ts
type FooAction = FooAction1 | FooAction2 | FooAction3

const isFooAction = (action: AnyAction) : action is FooAction => {
  return [ACTION_TYPE_1, ACTION_TYPE_2].includes(action.type); // buggy, missing `ACTION_TYPE_3` in array
}
```

The other part of type safety that I find greatly useful is exhaustive checking. To make that work in the switch I use this little util; 

```ts
export function exhaustiveAction<S, A extends Action>(
  state: S,
  _action: A & never
): S {
  return state;
}

...

function fooReducer(state = initialState, action: AnyAction) {
  if (isFooAction(action)) {
    switch(action.type) {
      case ...
      default:
        return exhaustiveAction(state, action)
    }
  }
  return state;
}
```

**Comment by phryneas** [comment]
Exhaustive checking is an illusion. Even if you could exhaustively check every action dispatched in your application in one reducer (let's face it, that would be a very long reducer, but you'd need to do that to be exhaustive, because every reducer is called with every action), you still would not check for the `INIT` action - that one even has a dynamic `type` string in development, because your reducer should explicitly not check for it. It's impossible to type that.

And at this point we don't even touch actions that might be used by middleware internally.

I can only recommend you get away from the concept of an "action union type" (that per definition will never be complete and always a lie to the compiler) and "exhaustive checking" (which as I stated above, is impossible) as soon as possible.

**Comment by MikelArnaiz** [comment]
Hey, @phryneas I get that, I read [your post about it](https://phryneas.de/redux-typescript-no-discriminating-union). I do not intend on covering all action in one reducer, and perhaps exhaustive checking is not an accurate name for what the example above does. It makes sure I don't forget to handle any action type but only those fulfilling the (parent) type guard.

Do you think that the little extra verbosity of the example above is not worth it? There isn't any lie with the typeguard, or is it just the idea of unions of actions what's wrong? Because to me the idea of not having any assurance that I'm covering all actions of a feature is worse.

How do you accomplice that in a non-RTK codebase? The answer might well be by leveraging this to the tests and not to Typescript.

**Comment by markerikson** [comment]
Closing as I think it's been discussed and there's nothing actionable for us here.
